Test/tutorial/Simple_compiler/Compiler.dfy

```diff
@@ -0,0 +1,498 @@
+  {
+    match e {
+      case Const(n) =>
+      case Op(op, e1, e2) =>
+        interpProg'_Concat(compileExpr(e2), compileExpr(e1), st);
+      // case Op(Sub, e1, e2) => // Here's the expanded version of the same proof
+      //   calc {
+      //     interpProg'(compileExpr(e), st);
+      //     interpProg'(Cons(PopSub, Concat(compileExpr(e2), compileExpr(e1))), st);
+      //     interpInstr(PopSub, interpProg'(Concat(compileExpr(e2), compileExpr(e1)), st));
+      //     { interpProg'_Concat(compileExpr(e2), compileExpr(e1), st); }
+      //   }
+    }
+  }
+
+  lemma compileStmtCorrect'(s: DafnyAST.Stmt, st: State)
+    ensures interpProg'(compileStmt(s), st) ==
+            State(st.stack, st.output + DafnyAST.interpStmt(s))
+  {
+    match s {
+      case Skip =>
+      case Print(e) =>
+        calc {
+          interpProg'(compileStmt(s), st);
+          interpProg'(compileStmt(DafnyAST.Print(e)), st);
+          interpProg'(Cons(PopPrint, compileExpr(e)), st);
+          interpInstr(PopPrint, interpProg'(compileExpr(e), st));
+          { compileExprCorrect'(e, st); }
+          interpInstr(PopPrint, State(Cons(DafnyAST.interpExpr(e), st.stack), st.output));
+        }
+      case Seq(s1, s2) =>
+        calc {
+          interpProg'(compileStmt(s), st);
+          interpProg'(compileStmt(DafnyAST.Seq(s1, s2)), st);
+          interpProg'(Concat(compileStmt(s2), compileStmt(s1)), st);
+          { interpProg'_Concat(compileStmt(s2), compileStmt(s1), st); }
+          interpProg'(compileStmt(s2), interpProg'(compileStmt(s1), st));
+        }
+    }
+  }
+
+  lemma compileStmtCorrect(s: DafnyAST.Stmt)
+    ensures interpProg(compileStmt(s)) == DafnyAST.interpStmt(s)
+  {
+    compileStmtCorrect'(s, EmptyState);
+  }
+}
+
+/// Connecting C# to Dafny
+/// ----------------------
+///
+/// Our Dafny ASTs are defined in terms of Dafny ``datatype`` values, but the AST produced using ANTLR on the C# side is defined through a class hierarchy (see ``Main.cs``).  To connect the two we need a translation function.  Here we write it in Dafny, which allows us to depend on fewer specifics of the compilation process.
+///
+/// To write the translator in Dafny we need a Dafny model of the C# class hierarchy.  Below we use traits, annotated with `{:extern}` to specify their C# names and `{:compile false}` to indicate that they should not be compiled: they act as shims that we can write the Dafny code against.  Then, once compiled to C#, our Dafny code can be linked against the real C# class hierarchy.
+///
+/// The model does not have to be completely faithful: for example we won't capture the fact that Expression and Statement share a base class on the C# side.
+///
+/// There are a few interesting difficulties: native C# types, nested types, enums:
+///
+/// 1. The C# side uses native C# types like `List<T>`.
+/// 2. C# types are commonly nested, but Dafny doesn't support nesting types.
```

Comment:
Could you link to https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/nested-types to clarify what a nested type is?